src/Interpreters/Cache/FileCache.cpp

```diff
@@ -202,8 +203,6 @@ void FileCache::initialize()
         keep_up_free_space_ratio_task = Context::getGlobalContextInstance()->getSchedulePool().createTask(log->name(), [this] { freeSpaceRatioKeepingThreadFunc(); });
         keep_up_free_space_ratio_task->schedule();
     }
-
-    is_initialized = true;
```

Comment:
We should not do `metadata.startup()` and `keep_up_free_space_ratio_task->schedule();` until metadata is loaded, otherwise data race or segfault is possible, because currently `loadMetadata` relies on the fact that cache is not accessed unless metadata is loaded, but the methods I mentioned could do cache access in the background threads which they launch. So probably we need to do all of initialize() inside a background thread, but FileStatus file should be created in advance in the main thread to prevent cache breakage.

Reply:
Thanks for the detailed comment, I'll look into this next week.
> currently loadMetadata relies on the fact that cache is not accessed unless metadata is loaded

Would changing this be enough to be able to use the cache even if partially initialized?

Reply:
> Would changing this be enough to be able to use the cache even if partially initialized?

No, currently code of loadMetadata relies on the fact that noone used cache until it is fully initialized. There are certain places where we do not take required mutexes because we know that it is initialization code. For example this one https://github.com/ClickHouse/ClickHouse/blob/3ef05fd11afb7e5827fda98cb833486649b340e4/src/Interpreters/Cache/FileCache.cpp#L1354
and also we cannot add new cache entries while cache is in the progress of initialization as the code of `loadMetadata` is not protected agains this

Reply:
> So probably we need to do all of initialize() inside a background thread, but FileStatus file should be created in advance in the main thread to prevent cache breakage.

Implemented these changes in https://github.com/ClickHouse/ClickHouse/pull/65736/commits/544a40c72a5cdaf63c3e820c83f68e36bd13ab24